Some principle for the filesystems.

paulp opened the pull request:
And the machinery to create lots more demonstrations.

paulp committed:
Some principle for the filesystems.

And the machinery to create lots more demonstrations.

EECOLOR commented on suffuse/node_fuse.go:
```diff
@@ -25,24 +24,16 @@ func (x *Elem) Access(ctx context.Context, req *f.AccessRequest) error {
-    subpath, cmd := splitSedSuffix(path)
-    if cmd == "" {
-      return err
+  for _, rule := range Rules {
```
Why are you skipping the first rule?

paulp replied:
The _ isn't the first rule, it's go's ridiculous iterator index which you can't avoid. It takes the values 0, 1, 2, ... while rule goes through the Rule values.

EECOLOR replied:
I must have read the documentation too fast and I missed the part in parentheses:

> If you only need the second item in the range (the value)

EECOLOR commented on suffuse/node_fuse.go:
```diff
@@ -25,24 +24,16 @@ func (x *Elem) Access(ctx context.Context, req *f.AccessRequest) error {
+  for _, rule := range Rules {
+    a := rule.MetaData(x.Path)
+    if a != nil {
+      *attr = *a
```
I have no experience with `*` so I leave it up to you to double check them

EECOLOR commented on suffuse/root.go:
```diff
@@ -0,0 +1,86 @@
+ *  when filesystems become more complex.
+ */
+type IdRule  struct { BaseRule }
+type SedRule struct { BaseRule }
```
I feel both `IdRule` and `SedRule` should be in a separate file. Maybe even in another directory, something like `rules`

paulp replied:
You can't move stuff into subdirectories without painful consequences. It's a different package.

EECOLOR replied:
I think rule implementations deserve their own package. Even for demo purposes that will be a place where we will add more. It's also nice for other people to be able to add them in a package that is less scary.

paulp replied:
I used to have more than one package. There is tremendous friction against multiple packages. You have to restate the entire path to the other packages in every file, e.g. if there's a "suffuse/rules" package then every file has to include

```
import "github.com/suffuse/go-suffuse/suffuse"
```

And every file has to now qualify every identifier from the suffuse package with a suffuse. prefix. If you try to use dot-imports you find if there's a single conflict you can't use dot imports at all because there's no facility for masking out identifiers from a package. (They fucked this up exactly the same way rust did, neither could be bothered to look at scala.) It goes on. I really think we should save the finer points for scala.

EECOLOR replied:
Ok, I see the pain in that, thanks for explaining.

Is there any specific reason the file is called `root.go` instead of `rules.go`?

dwijnand replied:
> You have to restate the entire path to the other packages in every file

Will go-suffuse one day be able to use suffuse to escape this tyranny?

paulp replied:
Yes, that's one of the comic aspects of this endeavor. Not even "one day" but like "next week" if we were sufficiently interested. It's why we should heavily prioritize shipping something stable, even if it does barely anything. Simply being able to organize our own source code is reason enough.

EECOLOR replied:
@paulp Could you comment on:

> Is there any specific reason the file is called root.go instead of rules.go?

paulp replied:
There's no reason, I'll rename it.

EECOLOR commented on suffuse/root.go:
```diff
@@ -0,0 +1,86 @@
+type Rule interface {
+  MetaData(Path)*fuse.Attr
+  FileData(Path)[]byte
+  DirData(Path)[]fuse.Dirent
```
If this `Rule` interface will stay here when the 'engine' is moved, I think we should abstract over the types from the `fuse` package.

paulp replied:
That gets into go-lang difficulties I haven't figured out how to overcome. Let's not work too hard trying to implement good abstractions in go, because you can't.

EECOLOR replied:
Of course I can't, I have been reading go for only 3 days, but you can ;-)

I'll take your word for it

EECOLOR commented on suffuse/root.go:
```diff
@@ -0,0 +1,86 @@
+ *  and then not have to implement irrelevant methods. There's
+ *  no default for Metadata because a filesystem which never
+ *  returns metadata is a useless filesystem.
+ */
```
We should add something like:

> If a rule does not match, it should return `nil`. This is especially important in case of the methods that return an array.

paulp replied:
At Rule it says "The fuse operations use the first non-nil return value found for a given Path," isn't that the same thing?

EECOLOR replied:
Crap, sorry about that, missed that. Ignore what I said.

EECOLOR commented on suffuse/root.go:
```diff
@@ -0,0 +1,86 @@
+ *  file, but for now it's hardcoded here.
+ */
+var Rules = []Rule {
+  IdRule{},  // The identity fs - passes underlying paths through unchanged
```
Shouldn't this be the last one in the array?

paulp replied:
There's no definitively correct order. Putting identity first means you can count on getting the actual file if there is a file with the requested name. Putting it last means any synthetic filesystem can intercept and supply something else. If there were filters, they would have to be in front of identity, but there aren't any filters yet.

EECOLOR replied:
Ahh, I see what you mean. Putting it first allows the other rules to take advantage of id's results.

paulp replied:
Not the way it's implemented now, no - it returns the first match, the other rules get no chance at it. This problem is far-reaching and cannot be solved here: filesystem composition is order dependent and there's no a priori correct way. I think we should remain focused on the behavior of individual filesystems and save composition problems for later. They are only bundled right now for convenience.

EECOLOR replied:
Yeah I realized that with comment about returning on first match.

I totally see your point and the inherent complexity of the problem. Agreed to leave it for now

EECOLOR commented on suffuse/node_fuse.go:
```diff
@@ -51,39 +42,53 @@ func (x *Elem) Lookup(ctx context.Context, name string) (fs.Node, error) {
+  path := x.Path
+  logD("ReadAll", "path", path)
+
+  for _, rule := range Rules {
```
You might want to consider moving the `for rule := range Rules` construct to a central location. That allows you to more safely handle rule mismatches.

paulp replied:
You mean and pass a Rule => T function to that? Ha ha, I think you need more go experience.

EECOLOR replied:
I certainly do, haha

EECOLOR commented:
I like the abstraction you created here!

paulp committed:
Move Rule into rule.go.

paulp committed:
Removed a bunch of abstraction.

Still trying to figure out go-lang well enough to get down
to the essence of things without unnecessary elaboration.
Like that sentence didn't.

Renamed "Elem" to "IdNode".

paulp committed:
More indirection removal.

Centralized all the global vars (which are mostly
"constants" to the extent that go-lang makes that possible,
which is to say, none) in suffuse.go.

EECOLOR commented on suffuse/node.go:
```diff
@@ -39,7 +21,7 @@ func HandleRead(req *f.ReadRequest, resp *f.ReadResponse, data []byte) {
 }
 
-func (x *Elem) String() string { return x.Path.Path }
+func (x *IdNode) String() string { return x.Path.Path }
```
My guess is that this function should be moved to the `inode.go` file

EECOLOR commented on suffuse/sfs.go:
```diff
@@ -34,7 +38,7 @@ func (u *Sfs) Destroy(ctx context.Context) {
 func (u *Sfs) Statfs(ctx context.Context, req *fuse.StatfsRequest, resp *fuse.StatfsResponse) error {
   logD("Statfs")
-  stat, err := u.RootNode.GetPath().SysStatfs()
+  stat, err := RootPath.SysStatfs()
```
I do not have a complete understanding of the semantics here, so I'll just ask. 

Does `u.RootNode.GetPath()` have the same semantics as `RootNode`?

From a naive perspective there is a difference between the Sfs RootNode and the absolute RootPath

paulp replied:
No, not at all. But neither of them provides meaningful information - I am forwarding statfs from the physical filesystem which happens to underlie a particular path. A fuse filesystem might be on top of multiple physical filesystems or on top of none. This is just to see that it's working.

EECOLOR committed:
Moved the 'String' method of 'IdNode' to the 'idnode.go' file
